src/network/networkd-brvlan.c

```diff
@@ -0,0 +1,330 @@
+
+        r = sd_rtnl_message_link_set_family(req, PF_BRIDGE);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not set message family: %m");
+
+
+        r = sd_netlink_message_open_container(req, IFLA_AF_SPEC);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not open IFLA_AF_SPEC container: %m");
+
+        // todo add flag self in case of bridge
+
+        // add vlan info
+        r = append_vlan_data(link, req, pvid, br_vlan_bitmap, br_untagged_bitmap);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not append VLANs: %m");
+
+        r = sd_netlink_message_close_container(req);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not close IFLA_AF_SPEC container: %m");
+
+        /* send message to the kernel */
+        r = sd_netlink_call_async(rtnl, req, set_brvlan_handler, link, 0, NULL);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not send rtnetlink message: %m");
+
+        return 0;
+}
+
+static inline void set_bit(unsigned nr, uint32_t *addr) {
+        if (nr < BRIDGE_VLAN_BITMAP_MAX)
+                addr[nr / 32] |= (((uint32_t) 1) << (nr % 32));
+}
+
+static inline int is_vlan_valid(unsigned vlan) {
+        if (vlan > VLANID_MAX)
+                return -EINVAL;
+        return 0;
+}
+
+static int parse_vid_range(const char *rvalue, int *vid, int *vid_end) {
+        int r;
+        char *p;
+        _cleanup_free_ char *_rvalue = NULL;
+
+        assert(rvalue);
+        assert(vid);
+        assert(vid_end);
+
+        _rvalue = strdup(rvalue);
+        if (!_rvalue)
+                return log_oom();
+
+        p = strchr(_rvalue, '-');
+        if (p) {
+                *p = '\0';
+                p++;
+                r = safe_atoi(_rvalue, vid);
+                if (0 == r)
+                        r = safe_atoi(p, vid_end);
+        } else {
+                r = safe_atoi(_rvalue, vid);
+        }
+        return r;
+}
+
+int config_parse_brvlan_vlan(const char *unit, const char *filename,
+                             unsigned line, const char *section,
+                             unsigned section_line, const char *lvalue,
+                             int ltype, const char *rvalue, void *data,
+                             void *userdata) {
+        Network *network = userdata;
+        int r;
+        int vid = -1;
+        int vid_end = -1;
+
+        assert(filename);
+        assert(section);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        r = parse_vid_range(rvalue, &vid, &vid_end);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "Failed to parse VLAN, ignoring: %s", rvalue);
+                return 0;
+        }
+
+        r = is_vlan_valid(vid);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Invalid VLAN ID %d", vid);
```

Comment:
yes, it does. The basic rule is: pass any errno you got, so that this ends up in the journal as structured metadata for the log message. If you don't have any errno, then pass 0. And in any case the error is really just about passing information along, it has no effect on whether something is fatal or not.